File: SortWare/Controls/TypeSelector.vb
Public Class TypeSelector

    Protected _allowedExtensions As String = ""

    Public Sub New()

        ' This call is required by the designer.
        InitializeComponent()

        ' Add any initialization after the InitializeComponent() call.

    End Sub
    Public Sub TypeSelect_Load() Handles Me.Load
        Dim itypes = Generics._imgExtensions.Split(" "c)
        Dim vtypes = Generics._vidExtensions.Split(" "c)
        Dim misctypes = Generics._miscExtensions.Split(" "c)
        For Each i In itypes
            TreeView1.Nodes.Item(0).Nodes.Add(i)
        Next
        For Each t In vtypes
            TreeView1.Nodes.Item(1).Nodes.Add(t)
        Next
        For Each m In misctypes
            TreeView1.Nodes.Item(2).Nodes.Add(m)
        Next
        'TreeView1.ExpandAll()
    End Sub

    Public Function isAllowed(ByVal Path As String) As Boolean
        Dim ext As String = IO.Path.GetExtension(Path).ToLower
        Return _allowedExtensions.ToLower.Contains(ext)
    End Function

    Private Sub TreeView1_AfterSelect(sender As Object, e As TreeViewEventArgs) Handles TreeView1.AfterCheck, TreeView1.AfterSelect
        RemoveHandler TreeView1.AfterCheck, AddressOf TreeView1_AfterSelect

        For Each node As TreeNode In e.Node.Nodes
            node.Checked = e.Node.Checked
        Next

        If e.Node.Checked Then
            If e.Node.Parent IsNot Nothing Then
                Dim allChecked As Boolean = True

                For Each node As TreeNode In e.Node.Parent.Nodes
                    If Not node.Checked Then
                        allChecked = False
                    End If
                Next

                If allChecked Then
                    e.Node.Parent.Checked = True
                End If
            End If
        Else
            If e.Node.Parent IsNot Nothing Then
                e.Node.Parent.Checked = False
            End If
        End If

        _allowedExtensions = ""
        For Each n As TreeNode In GetCheck(TreeView1.Nodes)
            _allowedExtensions &= n.Text.Replace("/", " ") & " "
        Next

        AddHandler TreeView1.AfterCheck, AddressOf TreeView1_AfterSelect
        RaiseEvent CheckChanged()
    End Sub

    Private Function GetCheck(ByVal node As TreeNodeCollection) As List(Of TreeNode)
        Dim lN As New List(Of TreeNode)
        For Each n As TreeNode In node
            If n.Checked AndAlso n.Tag Is Nothing Then
                lN.Add(n)
            End If
            lN.AddRange(GetCheck(n.Nodes))
        Next

        Return lN
    End Function

    Public Event CheckChanged()

End Class

File: SortWare/Controls/TypeSelector.cs
using System;
using System.Collections.Generic;
using System.Windows.Forms;

namespace SortWare
{
    public partial class TypeSelector
    {

        protected string _allowedExtensions = "";

        public TypeSelector()
        {

            // This call is required by the designer.
            InitializeComponent();

            // Add any initialization after the InitializeComponent() call.

        }
        public void TypeSelect_Load()
        {
            string[] itypes = Generics._imgExtensions.Split(' ');
            string[] vtypes = Generics._vidExtensions.Split(' ');
            string[] misctypes = Generics._miscExtensions.Split(' ');
            foreach (var i in itypes)
                TreeView1.Nodes[0].Nodes.Add(i);
            foreach (var t in vtypes)
                TreeView1.Nodes[1].Nodes.Add(t);
            foreach (var m in misctypes)
                TreeView1.Nodes[2].Nodes.Add(m);
            // TreeView1.ExpandAll()
        }

        public bool isAllowed(string Path)
        {
            string ext = System.IO.Path.GetExtension(Path).ToLower();
            return _allowedExtensions.ToLower().Contains(ext);
        }

        private void TreeView1_AfterSelect(object sender, TreeViewEventArgs e)
        {
            TreeView1.AfterCheck -= TreeView1_AfterSelect;

            foreach (TreeNode node in e.Node.Nodes)
                node.Checked = e.Node.Checked;

            if (e.Node.Checked)
            {
                if (e.Node.Parent is not null)
                {
                    bool allChecked = true;

                    foreach (TreeNode node in e.Node.Parent.Nodes)
                    {
                        if (!node.Checked)
                        {
                            allChecked = false;
                        }
                    }

                    if (allChecked)
                    {
                        e.Node.Parent.Checked = true;
                    }
                }
            }
            else if (e.Node.Parent is not null)
            {
                e.Node.Parent.Checked = false;
            }

            _allowedExtensions = "";
            foreach (TreeNode n in GetCheck(TreeView1.Nodes))
                _allowedExtensions += n.Text.Replace("/", " ") + " ";

            TreeView1.AfterCheck += TreeView1_AfterSelect;
            CheckChanged?.Invoke();
        }

        private List<TreeNode> GetCheck(TreeNodeCollection node)
        {
            var lN = new List<TreeNode>();
            foreach (TreeNode n in node)
            {
                if (n.Checked && n.Tag is null)
                {
                    lN.Add(n);
                }
                lN.AddRange(GetCheck(n.Nodes));
            }

            return lN;
        }

        public event CheckChangedEventHandler CheckChanged;

        public delegate void CheckChangedEventHandler();

        private void TypeSelect_Load(object sender, EventArgs e) => TypeSelect_Load();

    }
}